// @opennextjs/aws/dist/build/patch/astCodePatcher.js
// Mostly copied from the cloudflare adapter
import { readFileSync } from "node:fs";
import { Lang, parse, } from "@ast-grep/napi";
import yaml from "yaml";
/**
 * Returns the `Edit`s and `Match`es for an ast-grep rule in yaml format
 *
 * Notes:
 * - The rule must not contain a `transform` field
 * - The rule may contain a `fix` field, which is used to generate the edits.
 *   If not provided, no edits will be generated and the matches will be returned as is.
 * - You can use https://ast-grep.github.io/playground.html to create and verify rules.
 *
 * @param rule The rule. Either a yaml string or an instance of `RuleConfig`
 * @param root The root node
 * @param once only apply once
 * @returns A list of edits and a list of matches.
 */
export function applyRule(rule, root, { once = false } = {}) {
    const ruleConfig = typeof rule === "string" ? yaml.parse(rule) : rule;
    if (ruleConfig.transform) {
        throw new Error("transform is not supported");
    }
    const matches = once
        ? [root.find(ruleConfig)].filter((m) => m !== null)
        : root.findAll(ruleConfig);
    const edits = [];
    const fix = ruleConfig.fix;
    if (fix !== undefined) {
        matches.forEach((match) => {
            edits.push(match.replace(
            // Replace known placeholders by their value
            fix
                .replace(/\$\$\$([A-Z0-9_]+)/g, (_m, name) => match
                .getMultipleMatches(name)
                .map((n) => n.text())
                .join(""))
                .replace(/\$([A-Z0-9_]+)/g, (m, name) => match.getMatch(name)?.text() ?? m)));
        });
    }
    return { edits, matches };
}
/**
 * Parse a file and obtain its root.
 *
 * @param path The file path
 * @param lang The language to parse. Defaults to TypeScript.
 * @returns The root node for the file.
 */
export function parseFile(path, lang = Lang.TypeScript) {
    return parseCode(readFileSync(path, { encoding: "utf-8" }), lang);
}
/**
 * Parse a code string and obtain its root.
 *
 * @param code The code string
 * @param lang The language to parse. Defaults to TypeScript.
 * @returns The root node for the code string.
 */
export function parseCode(code, lang = Lang.TypeScript) {
    return parse(lang, code).root();
}
/**
 * Patches the code from by applying the rule.
 *
 * This function is mainly for on off edits and tests,
 * use `getRuleEdits` to apply multiple rules.
 *
 * @param code The source code
 * @param rule The astgrep rule (yaml or NapiConfig)
 * @param lang The language used by the source code
 * @param lang Whether to apply the rule only once
 * @returns The patched code
 */
export function patchCode(code, rule, { lang = Lang.TypeScript, once = false } = {}) {
    const node = parse(lang, code).root();
    const { edits } = applyRule(rule, node, { once });
    return node.commitEdits(edits);
}
/**
 * @param rule
 * @param lang
 * @returns a callback applying the the rule.
 */
export function createPatchCode(rule, lang = Lang.TypeScript) {
    return async ({ code }) => patchCode(code, rule, { lang });
}
